### PR TITLE
Add '--logging-format' flag to kube-controller-manager

### PR DESCRIPTION
**Pull request by SataQiu**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add `--logging-format` flag to kube-controller-manager.

**Which issue(s) this PR fixes**:

Ref kubernetes/enhancements#1602

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
kube-controller-manager: add '--logging-format' flag to support structured logging
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

### CODE REVIEW

**Comment by cheftako** [comment]
I have some of the same concerns as https://github.com/kubernetes/kubernetes/pull/91501#issuecomment-634932610.

Also a bit confused about the usage. It seems to be set up for dynamic configuration (use of locks etc on Set and Register and the fact that they are public) but the validate does not use locks and will only be checked start up. This seems prone to race conditions (except for things setup in the init setup).

**Comment by serathius** [comment]
Hey @cheftako 
We are working on implementing json format, hope to get it merged soon.

As for usage, flag was not implemented to be thread safe, internal register is, as we use it in a global variable. It was not implemented to support dynamic configuration, not I think we should support it. I'm not familiar with it, so I could be wrong.

I don't think there is a sense to make this flag thread safe. WDYT?

**Comment by serathius** [comment]
Json format was added. https://github.com/kubernetes/kubernetes/issues/91490
PR to update help listing available options is in flight. https://github.com/kubernetes/kubernetes/pull/92177

I think we can proceed with this PR. @lavalamp PTAL

**Comment by cheftako** [comment]
> As for usage, flag was not implemented to be thread safe, internal register is, as we use it in a global variable. It was not implemented to support dynamic configuration, not I think we should support it. I'm not familiar with it, so I could be wrong.
> 
> I don't think there is a sense to make this flag thread safe. WDYT?

I agree. I understand that your usage is not intended to be thread safe. However the logs library seems confused on this. It has a Register() method, whish is public, uses lock to ensure safe addition on new loggers and the documentation that does not mention any restrictions on adding logger methods. To me this implies new loggers can be added at any time. However the init() and Validate() methods in it suggest, as you say, that it does not support dynamic configuration. 

Lets take a look once we have Json format.

**Comment by serathius** [comment]
PR updating flag to inform about JSON foramt was merged. Can go move forward with this one? @cheftako

**Comment by serathius** [comment]
/hold
Reason same as https://github.com/kubernetes/kubernetes/pull/91522#issuecomment-647622039

**Comment by serathius** [comment]
/unhold 
Flag validation was implemented, https://github.com/kubernetes/kubernetes/pull/92394#issuecomment-652046227
/assign @sttts

File: cmd/kube-controller-manager/app/options/options.go [hunk 1]
```diff
@@ -249,6 +252,7 @@ func (s *KubeControllerManagerOptions) Flags(allControllers []string, disabledBy
 	s.SAController.AddFlags(fss.FlagSet("serviceaccount controller"))
 	s.TTLAfterFinishedController.AddFlags(fss.FlagSet("ttl-after-finished controller"))
 	s.Metrics.AddFlags(fss.FlagSet("metrics"))
+	s.Logs.AddFlags(fss.FlagSet("logs"))
```
Comment (sttts) [inline cmd/kube-controller-manager/app/options/options.go:255]: Does this fit better into generic? Is it just one flag? Where are the other klog flags?

Comment (serathius) [reply cmd/kube-controller-manager/app/options/options.go:255]: Othe klog flags are Global flags. I think this flag should be in same flagSet `logs`  as in other control plane components.

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:255]: https://gist.github.com/fd93b5d6044d7c256d2e70fcabf503ae does it. All logging flags will be under the logs section.

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:255]: My gist does not move `log-flush-frequency` yet. That's also defined in component-base/logs and has to go to the named flag set.

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:255]: As the proposed change is bigger (removes the init func behaviour of the package), I am fine to postpone this to 1.20. But would like to see a PR started with the gist, tagged with the relevant people **before** approving this PR here.

Comment (serathius) [reply cmd/kube-controller-manager/app/options/options.go:255]: Done https://github.com/kubernetes/kubernetes/pull/92707

Comment (sttts) [reply cmd/kube-controller-manager/app/options/options.go:255]: We agreed to keep this new logs section for now, and move klog flags over in a follow-up (in 1.19 or 1.20 if we miss code freeze).

**Pull request merged**